api/all/src/main/java/io/opentelemetry/api/common/ArrayBackedAttributes.java

```diff
@@ -26,6 +27,75 @@ private ArrayBackedAttributes(Object[] data, Comparator<AttributeKey<?>> keyComp
     super(data, keyComparator);
   }
 
+  /** Default impelmentation for remove all method added to Attributes. */
+  static Attributes removeAllImpl(Attributes source, Attributes toRemove) {
+    if (toRemove.isEmpty()) {
+      return source;
+    }
+    if (source instanceof ArrayBackedAttributes && toRemove instanceof ArrayBackedAttributes) {
```

Comment:
It does mean though that maybe we don't need to add `removeAll` to the public API - can we think of any use case besides exemplar? It might be better to keep the slow implementation only for now, private to metrics, and reconsider how to optimize it, taking into account the instrumentation-api implementation too

Reply:
Note that I sort of wanted `UnsafeAttributes` to be made a public class in the `internal` package in this repo. One reason was to reduce dupe with https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/AttributesMap.java though I feel as if I had an even better reason that I can't recall right now... But now I have another one :) to allow adding the `removeAll` optimization without necessarily adding it to the public API if we don't see a use outside of exemplars.

Reply:
If you'd like to hide this in the SDK, I guess we can make that happen.  I suspect we'll find it useful for more than just metrics at some point (e.g. removing resource attribtues from trace attributes, or trace attributes from event attribtues).

I guess a few important points:

1. I need access to UnsafeAttributes in the metrics SDK if you want to make it an implementation detail.
2. UnsafeAttributes is also implemented in the Trace SDK in similar fashion.
3. For UnsafeAttributes the efficiency is actually acceptable in the "slow" path because you have O(1) lookup.  It's only Array-based against Array-based that's an issue.  Probably an early optimization, but also an easy one.
4. We use ArrayBackedAttributes in all the unit tests in the SDK *and* benchmarking test, where this performance tweak can show up.
5. Attributes is an exposed implementation that anyone could define how they want.

Reply:
I realized because this API is significantly new functionality, not a simple shortcut or a Java-ism, we should get it in the `opentelemetry-specification` first before adding it and possibly diverging on this non-alpha artifact. So let's start with keeping this method hidden, and moving stuff to `internal` as needed to allow that

Reply:
New process is kind of meant to be the inverse.  We first expose something in an "ok to change" way, then use its prototype to justify a specification change.  I.e. there's a tension between "bottom up" and "top down" specification now.

I'm closing this and migrating to the SDK internal-only option.